File: packages/typescript/src/modules/payments.ts
/**
 * PaymentsModule: process, pre-authorize, capture, void, hosted checkout.
 */

import { mapReturnCodeToError, ValidationException } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { BASE_URLS } from '../config.js';
import type {
  CreateTransactionRequest,
  CreateTransactionResponse,
  PaymentInfoItem,
  SubserviceItem,
} from '../types/api.js';
import type { PaymentIntent, TransactionResult } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';
import { validatePaymentIntent } from '../utils/validators.js';

/** States where a double-payment retry must be blocked */
const DOUBLE_PAYMENT_STATES = ['BANK', 'PENDING', 'CAPTURED', 'CREATED'];

function buildReferenceArray(intent: PaymentIntent): string[] {
  const c = intent.customer;
  const refs: string[] = [
    c.documentType ?? '',
    c.documentNumber ?? '',
    intent.merchantTransactionId ?? '',
    c.fullName ?? '',
    c.email ?? '',
    c.phone ?? '',
  ];
  if (intent.additionalReferences) {
    refs.push(...intent.additionalReferences);
  }
  return refs;
}

function info(code: number, desc: string, value: string | number | undefined): PaymentInfoItem | null {
  if (value === undefined || value === null || value === '') return null;
  return { AttributeCode: code, AttributeDesc: desc, AttributeValue: String(value) };
}

function buildPaymentInfoArray(intent: PaymentIntent): PaymentInfoItem[] {
  const items: Array<PaymentInfoItem | null> = [
    info(6, 'Usermail', intent.customer.email),
    info(17, 'CardHolderName', intent.customer.fullName),
    info(18, 'CardHolderIdType', intent.customer.documentType),
    info(19, 'CardHolderId', intent.customer.documentNumber),
    info(23, 'IPAddress', intent.ipAddress),
    info(24, 'DeviceFingerPrint', intent.deviceFingerPrint),
    info(25, 'OneTimePassword', intent.oneTimePassword),
    info(26, 'MerchantTransactionId', intent.merchantTransactionId),
    info(34, 'UserType', intent.userType),
  ];

  // Extra paymentInfo map
  if (intent.paymentInfo) {
    for (const [desc, value] of Object.entries(intent.paymentInfo)) {
      // For known descriptors map to code; otherwise skip (desc is used as desc)
      items.push({ AttributeCode: 0, AttributeDesc: desc, AttributeValue: value });
    }
  }

  return items.filter((x): x is PaymentInfoItem => x !== null);
}

function buildTokenInfoArray(intent: PaymentIntent): PaymentInfoItem[] | undefined {
  if (!intent.tokenId) return undefined;

  const items: Array<PaymentInfoItem | null> = [
    info(1, 'TokenId', intent.tokenId),
    info(2, 'PaymentSystem', intent.paymentSystem),
    info(3, 'SecureCode', intent.secureCode),
    info(5, 'Installments', intent.installments),
    info(6, 'Usermail', intent.customer.email),
    info(9, 'FiCode', intent.fiCode),
    info(19, 'CardHolderId', intent.customer.documentNumber),
    info(25, 'OneTimePassword', intent.oneTimePassword),
  ];

  return items.filter((x): x is PaymentInfoItem => x !== null);
}

function buildSubservices(intent: PaymentIntent): SubserviceItem[] | undefined {
  if (!intent.subservices || intent.subservices.length === 0) return undefined;
  return intent.subservices.map((s) => ({
    EntityCode: s.entityCode,
    SrvCode: String(s.srvCode),
    ValueType: s.valueType,
    TransValue: s.transValue,
    TransVatValue: s.transVatValue,
  }));
}

function mapResponse(res: CreateTransactionResponse): TransactionResult {
  const tr = res.TransactionResponse;
  return {
    returnCode: res.ReturnCode,
    ticketId: res.TicketId ?? tr?.TicketId,
    eCollectUrl: res.eCollectUrl,
    lifetimeSecs: res.LifetimeSecs,
    tranState: (tr?.TranState as TransactionResult['tranState']) ?? undefined,
    trazabilityCode: tr?.TrazabilityCode,
    transValue: tr?.TransValue,
    transVatValue: tr?.TransVatValue,
    payCurrency: tr?.PayCurrency,
    currencyRate: tr?.CurrencyRate,
    bankProcessDate: tr?.BankProcessDate,
    fiCode: tr?.FICode,
    fiName: tr?.FiName,
    paymentSystem: tr?.PaymentSystem,
    tranCycle: tr?.TransCycle,
    invoice: tr?.Invoice,
    referenceArray: tr?.ReferenceArray,
    srvCode: tr?.SrvCode,
    paymentInfoArray: tr?.PaymentInfoArray?.map((i) => ({
      code: i.AttributeCode,
      desc: i.AttributeDesc,
      value: i.AttributeValue,
    })),
  };
}

export class PaymentsModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
  }

  private get baseUrl(): string {
    return BASE_URLS[this.config.environment];
  }

  private async _send(
    body: Omit<CreateTransactionRequest, 'SessionToken'>,
  ): Promise<TransactionResult> {
    const url = `${this.baseUrl}/createTransactionPayment`;
    const sessionToken = await this.session.getActive();

    const fullBody: CreateTransactionRequest = { ...body, SessionToken: sessionToken };

    const res = await this.http.post<CreateTransactionRequest, CreateTransactionResponse>(
      url,
      fullBody,
    );

    if (res.ReturnCode === 'FAIL_APIEXPIREDSESSION') {
      this.session.invalidate();
      fullBody.SessionToken = await this.session.getActive();
      const retried = await this.http.post<CreateTransactionRequest, CreateTransactionResponse>(
        url,
        fullBody,
      );
      if (retried.ReturnCode !== 'SUCCESS') {
        throw mapReturnCodeToError(retried.ReturnCode);
      }
      return mapResponse(retried);
    }

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return mapResponse(res);
  }

  /**
   * Process a payment (RequestType=0 — immediate authorization).
   */
  async process(intent: PaymentIntent): Promise<TransactionResult> {
    validatePaymentIntent(intent, this.config.etyCode);

    const srvCode = intent.srvCode ?? this.config.srvCode;
    if (!srvCode) {
      throw new ValidationException('srvCode is required (set in PaymentIntent or EcollectClient config)');
    }

    return this._send({
      EntityCode: this.config.etyCode,
      SrvCode: srvCode,
      TransValue: intent.amount,
      TransVatValue: intent.vatAmount,
      SrvCurrency: intent.currency,
      URLRedirect: intent.redirectUrl,
      URLResponse: intent.responseUrl,
      LangCode: intent.langCode ?? 'ES',
      PaymentSystem: intent.paymentSystem,
      FICode: intent.fiCode,
      Invoice: intent.invoice,
      InvoiceDueDate: intent.invoiceDueDate,
      PolicyCode: intent.policyCode,
      RequestType: 0,
      ReferenceArray: buildReferenceArray(intent),
      SubservicesArray: buildSubservices(intent),
      PaymentInfoArray: buildPaymentInfoArray(intent),
      TokenInfoArray: buildTokenInfoArray(intent),
    });
  }

  /**
   * Pre-authorize a payment (RequestType=1 — reserve funds).
   */
  async preAuthorize(intent: PaymentIntent): Promise<TransactionResult> {
    validatePaymentIntent(intent, this.config.etyCode);

    const srvCode = intent.srvCode ?? this.config.srvCode;
    if (!srvCode) {
      throw new ValidationException('srvCode is required');
    }

    return this._send({
      EntityCode: this.config.etyCode,
      SrvCode: srvCode,
      TransValue: intent.amount,
      TransVatValue: intent.vatAmount,
      SrvCurrency: intent.currency,
      URLRedirect: intent.redirectUrl,
      URLResponse: intent.responseUrl,
      LangCode: intent.langCode ?? 'ES',
      PaymentSystem: intent.paymentSystem,
      FICode: intent.fiCode,
      PolicyCode: intent.policyCode,
      RequestType: 1,
      ReferenceArray: buildReferenceArray(intent),
      SubservicesArray: buildSubservices(intent),
      PaymentInfoArray: buildPaymentInfoArray(intent),
      TokenInfoArray: buildTokenInfoArray(intent),
    });
  }

  /**
   * Capture a pre-authorized payment (RequestType = ticketId positive).
   */
  async capture(ticketId: number, finalAmount?: number): Promise<TransactionResult> {
    if (!ticketId || ticketId <= 0) {
      throw new ValidationException('ticketId must be a positive number');
    }

    return this._send({
      EntityCode: this.config.etyCode,
      RequestType: ticketId,
      TransValue: finalAmount,
    });
  }

  /**
   * Void a pre-authorized payment (RequestType = -ticketId).
   */
  async void(ticketId: number): Promise<TransactionResult> {
    if (!ticketId || ticketId <= 0) {
      throw new ValidationException('ticketId must be a positive number');
    }

    return this._send({
      EntityCode: this.config.etyCode,
      RequestType: -ticketId,
    });
  }

  /**
   * Hosted checkout: redirect user to ecollect payment page.
   * Returns the eCollectUrl for redirect.
   */
  async hostedCheckout(intent: PaymentIntent): Promise<TransactionResult> {
    if (!intent.redirectUrl) {
      throw new ValidationException('redirectUrl is required for hostedCheckout');
    }
    return this.process(intent);
  }

  /**
   * Process with additional PaymentInfoItems injected (used by PaymentLinksModule).
   * @internal
   */
  async _processWithExtraInfo(
    intent: PaymentIntent,
    extraInfo: PaymentInfoItem[],
  ): Promise<TransactionResult> {
    validatePaymentIntent(intent, this.config.etyCode);

    const srvCode = intent.srvCode ?? this.config.srvCode;
    if (!srvCode) {
      throw new ValidationException('srvCode is required');
    }

    const paymentInfoArray = [...buildPaymentInfoArray(intent), ...extraInfo];

    return this._send({
      EntityCode: this.config.etyCode,
      SrvCode: srvCode,
      TransValue: intent.amount,
      TransVatValue: intent.vatAmount,
      SrvCurrency: intent.currency,
      URLRedirect: intent.redirectUrl,
      URLResponse: intent.responseUrl,
      LangCode: intent.langCode ?? 'ES',
      PaymentSystem: intent.paymentSystem,
      FICode: intent.fiCode,
      Invoice: intent.invoice,
      InvoiceDueDate: intent.invoiceDueDate,
      PolicyCode: intent.policyCode,
      RequestType: 0,
      ReferenceArray: buildReferenceArray(intent),
      SubservicesArray: buildSubservices(intent),
      PaymentInfoArray: paymentInfoArray,
      TokenInfoArray: buildTokenInfoArray(intent),
    });
  }

  /**
   * Check if a transaction state prevents retry (double-payment guard).
   */
  static isDoublePaymentState(tranState: string): boolean {
    return DOUBLE_PAYMENT_STATES.includes(tranState);
  }
}

File: packages/python/ecollect/modules/payments.py
"""Payment processing module."""
import logging
from typing import Any, Dict, List, Optional

from ecollect.config import EcollectConfig
from ecollect.exceptions import SessionExpiredError, raise_for_return_code
from ecollect.types import PaymentIntent
from ecollect.utils.http import AsyncHttpClient, run_sync
from ecollect.utils.validators import validate_payment_intent

logger = logging.getLogger(__name__)

# AttributeCode constants
_AC = {
    "CardNumber": 0,
    "TokenId": 1,
    "PaymentSystem": 2,
    "SecureCode": 3,
    "ExpirationDate": 4,
    "Installments": 5,
    "Usermail": 6,
    "MobileCountryCode": 7,
    "MobileNumber": 8,
    "FiCode": 9,
    "CardHolderName": 17,
    "CardHolderIdType": 18,
    "CardHolderId": 19,
    "AccountType": 22,
    "IPAddress": 23,
    "DeviceFingerPrint": 24,
    "OneTimePassword": 25,
    "MerchantTransactionId": 26,
    "UserType": 34,
    "LifetimeSecs": 35,
}


def _info_item(code: int, desc: str, value: str) -> Dict[str, Any]:
    return {"AttributeCode": code, "AttributeDesc": desc, "AttributeValue": str(value)}


def _build_reference_array(intent: PaymentIntent) -> List[str]:
    """Build ReferenceArray from a PaymentIntent following the documented order."""
    c = intent.customer
    return [
        c.document_type or "",
        c.document_number or "",
        intent.merchant_transaction_id or "",
        c.full_name or "",
        c.email or "",
        c.phone or "",
    ]


def _build_payment_info_array(intent: PaymentIntent) -> List[Dict[str, Any]]:
    """Build PaymentInfoArray from a PaymentIntent."""
    items: List[Dict[str, Any]] = []

    if intent.payment_system is not None:
        items.append(_info_item(_AC["PaymentSystem"], "PaymentSystem", str(intent.payment_system)))
    if intent.customer.email:
        items.append(_info_item(_AC["Usermail"], "Usermail", intent.customer.email))
    if intent.customer.document_type:
        items.append(_info_item(_AC["CardHolderIdType"], "CardHolderIdType", intent.customer.document_type))
    if intent.customer.document_number:
        items.append(_info_item(_AC["CardHolderId"], "CardHolderId", intent.customer.document_number))
    if intent.ip_address:
        items.append(_info_item(_AC["IPAddress"], "IPAddress", intent.ip_address))
    if intent.device_fingerprint:
        items.append(_info_item(_AC["DeviceFingerPrint"], "DeviceFingerPrint", intent.device_fingerprint))
    if intent.merchant_transaction_id:
        items.append(_info_item(_AC["MerchantTransactionId"], "MerchantTransactionId", intent.merchant_transaction_id))
    if intent.user_type is not None:
        items.append(_info_item(_AC["UserType"], "UserType", str(intent.user_type)))
    if intent.account_type is not None:
        items.append(_info_item(_AC["AccountType"], "AccountType", str(intent.account_type)))
    if intent.one_time_password:
        items.append(_info_item(_AC["OneTimePassword"], "OneTimePassword", intent.one_time_password))

    # Append any extra items provided by the caller
    items.extend(intent.extra_payment_info)
    return items


def _build_token_info_array(intent: PaymentIntent) -> List[Dict[str, Any]]:
    """Build TokenInfoArray from a PaymentIntent."""
    items: List[Dict[str, Any]] = []
    if intent.token_id:
        items.append(_info_item(_AC["TokenId"], "TokenId", intent.token_id))
    if intent.payment_system is not None:
        items.append(_info_item(_AC["PaymentSystem"], "PaymentSystem", str(intent.payment_system)))
    if intent.fi_code:
        items.append(_info_item(_AC["FiCode"], "FiCode", intent.fi_code))
    if intent.secure_code:
        items.append(_info_item(_AC["SecureCode"], "SecureCode", intent.secure_code))
    if intent.installments is not None:
        items.append(_info_item(_AC["Installments"], "Installments", str(intent.installments)))
    if intent.customer.email:
        items.append(_info_item(_AC["Usermail"], "Usermail", intent.customer.email))
    if intent.customer.document_number:
        items.append(_info_item(_AC["CardHolderId"], "CardHolderId", intent.customer.document_number))
    return items


class PaymentsModule:
    """Handles payment creation, pre-authorization, capture, void, and hosted checkout."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session

    async def _post(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Post to endpoint, auto-refreshing session on FAIL_APIEXPIREDSESSION."""
        url = self._config.endpoint_url(endpoint)
        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        if return_code == "FAIL_APIEXPIREDSESSION":
            logger.info("Session expired; refreshing and retrying once")
            self._session.invalidate()
            payload["SessionToken"] = await self._session.get_active()
            data = await self._http.post_with_retry(url, payload)
        raise_for_return_code(data.get("ReturnCode", ""), str(data))
        return data

    def _resolve_codes(self, intent: PaymentIntent) -> tuple:
        ety = intent.ety_code if intent.ety_code is not None else self._config.ety_code
        srv = intent.srv_code if intent.srv_code is not None else self._config.srv_code
        return ety, srv

    async def _build_payload(self, intent: PaymentIntent, request_type: Any = None) -> Dict[str, Any]:
        ety_code, srv_code = self._resolve_codes(intent)
        session_token = await self._session.get_active()

        payload: Dict[str, Any] = {
            "EntityCode": ety_code,
            "SessionToken": session_token,
            "TransValue": intent.amount,
            "SrvCurrency": intent.currency,
            "ReferenceArray": _build_reference_array(intent),
            "LangCode": intent.lang_code,
        }
        if srv_code is not None:
            payload["SrvCode"] = srv_code
        if intent.vat_value is not None:
            payload["TransVatValue"] = intent.vat_value
        if intent.url_redirect:
            payload["URLRedirect"] = intent.url_redirect
        if intent.url_response:
            payload["URLResponse"] = intent.url_response
        if intent.payment_system is not None:
            payload["PaymentSystem"] = intent.payment_system
        if intent.fi_code:
            payload["FICode"] = intent.fi_code
        if intent.invoice:
            payload["Invoice"] = intent.invoice
        if intent.invoice_due_date:
            payload["InvoiceDueDate"] = intent.invoice_due_date
        if intent.policy_code:
            payload["PolicyCode"] = intent.policy_code

        rt = request_type if request_type is not None else (intent.request_type if intent.request_type is not None else 0)
        payload["RequestType"] = rt

        payment_info = _build_payment_info_array(intent)
        if payment_info:
            payload["PaymentInfoArray"] = payment_info

        token_info = _build_token_info_array(intent)
        if token_info:
            payload["TokenInfoArray"] = token_info

        if intent.subservices:
            payload["SubservicesArray"] = intent.subservices

        return payload

    # ------------------------------------------------------------------
    # Async API
    # ------------------------------------------------------------------

    async def process(self, intent: PaymentIntent) -> Dict[str, Any]:
        """Process an immediate payment (RequestType=0)."""
        validate_payment_intent(intent)
        payload = await self._build_payload(intent, request_type=0)
        return await self._post("createTransactionPayment", payload)

    async def pre_authorize(self, intent: PaymentIntent) -> Dict[str, Any]:
        """Send a pre-authorization request (RequestType=1)."""
        validate_payment_intent(intent)
        payload = await self._build_payload(intent, request_type=1)
        return await self._post("createTransactionPayment", payload)

    async def capture(self, ticket_id: int, amount: Optional[float] = None, intent: Optional[PaymentIntent] = None) -> Dict[str, Any]:
        """Capture / post a pre-authorized transaction (RequestType=TicketId)."""
        session_token = await self._session.get_active()
        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "RequestType": ticket_id,  # positive TicketId = post
        }
        if amount is not None:
            payload["TransValue"] = amount
        if intent is not None:
            payload["SrvCurrency"] = intent.currency
            ety, srv = self._resolve_codes(intent)
            payload["EntityCode"] = ety
            if srv:
                payload["SrvCode"] = srv
        return await self._post("createTransactionPayment", payload)

    async def void_payment(self, ticket_id: int) -> Dict[str, Any]:
        """Void / reverse a pre-authorization (RequestType=-TicketId)."""
        session_token = await self._session.get_active()
        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "RequestType": -ticket_id,  # negative = anular
        }
        return await self._post("createTransactionPayment", payload)

    async def hosted_checkout(self, intent: PaymentIntent) -> Dict[str, Any]:
        """Create a hosted checkout — returns eCollectUrl for user redirection."""
        validate_payment_intent(intent)
        payload = await self._build_payload(intent, request_type=0)
        return await self._post("createTransactionPayment", payload)

    # ------------------------------------------------------------------
    # Sync wrappers
    # ------------------------------------------------------------------

    def process_sync(self, intent: PaymentIntent) -> Dict[str, Any]:
        return run_sync(self.process(intent))

    def pre_authorize_sync(self, intent: PaymentIntent) -> Dict[str, Any]:
        return run_sync(self.pre_authorize(intent))

    def capture_sync(self, ticket_id: int, amount: Optional[float] = None, intent: Optional[PaymentIntent] = None) -> Dict[str, Any]:
        return run_sync(self.capture(ticket_id, amount, intent))

    def void_payment_sync(self, ticket_id: int) -> Dict[str, Any]:
        return run_sync(self.void_payment(ticket_id))

    def hosted_checkout_sync(self, intent: PaymentIntent) -> Dict[str, Any]:
        return run_sync(self.hosted_checkout(intent))
